src/dune/context.ml

```diff
@@ -264,6 +264,49 @@ let check_fdo_support has_native ocfg ~name =
             version_string
         ]
 
+module DB = struct
+  module Settings = struct
+    let t = Fdecl.create Dyn.Encoder.opaque
+
+    let equal (e1, w1) (e2, w2) = Env.equal e1 e2 && Workspace.equal w1 w2
+
+    type state =
+      | Already_initialized
+      | Needs_initialization
+
+    let set ~env workspace =
+      match Fdecl.peek t with
+      | Some v ->
+        if equal v (env, workspace) then
```

Comment:
Why do we need this check? I thought the expectation was that `set` would be called only once

Reply:
I think this is slightly better. Now if a user enters a comment into a workspace field then we will not error out.

Reply:
Oh, my thinking was that if the user modified the workspace file while running in polling mode we would simply not reload it at all. I.e. you'd need to restart Dune.

Saying that, that just gave me a idea of how to do this another way. Let's say we don't do anything special and make `Context.get` a plain memoised function. But instead of exposing `get` to the user, we expose this wrapper instead:

```ocaml
let get name = Memo.peek_exn get name
```

Then during the initialisation phase, i.e. at the time we would normally call `set` we instead call `Memo.exec get name` to ensure the given memory cell is filled.

Reply:
What do you think?

Reply:
> Oh, my thinking was that if the user modified the workspace file while running in polling mode we would simply not reload it at all. I.e. you'd need to restart Dune.

To me this behavior isn't not optimal. This limitation isn't really intuitive and I think it's just better to notify the user somehow. So I think it's better to simply terminate dune when the user changes the workspace file.

> What do you think?

Let me give this a try. Truthfully, the advantage isn't obvious to me. But I don't see why this is worse either.

Reply:
The main advantage is that we should get the dependency tracking and sharing for free. So when the user modifies the config file, Dune would actually reload it

Reply:
I gave this a try and there are some problems with this approach. For example, what happens when a user removes a context from a workspace file? How do we make sure that the subsequent call to `Memo.peek_exn` will fail?

Reply:
`peek_exn` records the call just as `get`. So when the workspace config file changes, we'll know that the result of `peek_exn xxx` is no longer valid.

Reply:
So for `get` to be a memoization function, we're going to need a memoized version of `create`:

```ocaml
val create : env:Env.t -> Workspace.t -> t list Fiber.t
```

I'm wondering if we should go a bit deeper and have a database for workspaces then. Having this function simply call `Workspace.DB.get` seems natural to me.

Reply:
Actually, I don't yet see how this is going to work. To have this property:

> So when the workspace config file changes, we'll know that the result of peek_exn xxx is no longer valid.

It's enough to have this `get` function call some some memoized function `Workspace.DB.get` to get the defined workspace, and then a memoized `Context.create` to construct the contexts from it. However, I'm not sure how we'd avoid making it asynchronous in this case.

Reply:
> I'm wondering if we should go a bit deeper and have a database for workspaces then. Having this function simply call Workspace.DB.get seems natural to me.

> Do we need a full blown database or would using a memoized function be enough?
It's enough to have this get function call some some memoized function Workspace.DB.get to get the defined workspace, and then a memoized Context.create to construct the contexts from it. However, I'm not sure how we'd avoid making it asynchronous in this case.

We avoid the `Fiber.t` by calling `Memo.peek_exn`. As long as we make sure `Memo.exec` has been called between the time the file changed and the time we call `peek_exn`, we are sure that `peek_exn` will succeed.

Reply:
> Do you mean memoise Workspace.DB.get rather than a database?

Yeah, that is what I meant. Essentially, constructing the workspace only requires the following input:

```
type t =
  | Default of Context_name.t option * Profile.t option
  | Definition of Context_name.t option * Profile.t option * Path.t
```

I suppose we can also depend on the current run to make sure that we re-read the workspace file.

Reply:
Seems good to me!